enterprise/backend/src/metabase_enterprise/gsheets.clj

```diff
@@ -0,0 +1,350 @@
+                   [:status                      [:enum "not-connected" "loading" "complete"]]
+                   [:folder_url {:optional true} ms/NonBlankString]])
+
+(defn error-response-in-body
+  "We've decided to return errors as a map with an `:error` key. This function is a helper to make that map.
+
+  This formats and throws an ex-info that will put the message into the body of the response."
+  ([message] (error-response-in-body message {}))
+  ([message data]
+   (throw (ex-info message (merge data {:errors true :message message})))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; MB <-> HM APIs
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(mu/defn- hm-service-account-email :- [:or [:= false] :string]
+  "Checks to see if Google service-account is setup in harbormaster, and returns the email."
+  []
+  (let [[_status {:keys [body] :as _response}] (hm.client/make-request :get "/api/v2/mb/connections-google/service-account")]
+    (if-let [email (:email body)]
+      email
+      (error-response-in-body
+       (tru "Google service-account is not setup in harbormaster.")
+       {:status-code 502}))))
+
+(mr/def ::gdrive-conn [:map {:description "The Harbormaster Gdrive Connection"}
+                       [:id :string]
+                       [:type [:= "gdrive"]]
+                       [:status [:enum "initializing" "syncing" "active" "error"]]
+                       [:last-sync-at [:maybe :time/zoned-date-time]]
+                       [:last-sync-started-at [:maybe :time/zoned-date-time]]
+                       [:created-at :time/zoned-date-time]
+                       [:updated-at :time/zoned-date-time]])
+
+(defn- is-gdrive?
+  "Is this connection a gdrive connection?"
+  [{:keys [type] :as _conn}] (= "gdrive" type))
+
+(defn- normalize-gdrive-conn
+  "Normalize the gdrive connection shape from harbormaster, mostly parsing times."
+  [gdc]
+  (-> gdc
+      (dissoc :hosted-instance-resource)
+      (m/update-existing :last-sync-at u.date/parse)
+      (m/update-existing :last-sync-started-at u.date/parse)
+      (m/update-existing :created-at u.date/parse)
+      (m/update-existing :updated-at u.date/parse)))
+
+(mu/defn- hm-get-gdrive-conns :- [:sequential ::gdrive-conn]
+  "Get the harbormaster gdrive type connection.
+
+  If the response fails:    return an empty list.
+  If the response succeeds: return a list of all gdrive-type connections."
+  []
+  (let [[status response] (hm.client/make-request :get "/api/v2/mb/connections")]
+    (if (= status :ok)
+      (mapv normalize-gdrive-conn
+            (filter is-gdrive? (or (some-> response :body) [])))
+      [])))
+
+(mu/defn- hm-delete-conn :- ::hm.client/http-reply
+  "Delete (presumably a gdrive) connection on HM."
+  [conn-id]
+  (hm.client/make-request :delete (str "/api/v2/mb/connections/" conn-id)))
+
+(defn- reset-gsheets-status []
+  (gsheets! not-connected)
+  (doseq [{:keys [id]} (hm-get-gdrive-conns)]
+    (hm-delete-conn id))
```

Comment:
What happens here if these get out of sync? i.e. we reset the gsheets state in Metabase but HM fails to delete all the connections? Is that a recoverable state?

Reply:
That is the only way we will leak connections. It is still a possibility if hm fails to delete it, but I think this is good enough.

I am deleting all gdrive conns (though we use 1 and only 1) so that we don't leak connections.

If a user is disconnecting/reconnecting their drive folders then every time they disconnect (or hit an error), we try to clean up orphanted gdrive conns.